tests/appsTest.js

```diff
@@ -15,10 +15,133 @@ fdescribe('Main: Currently testing apps management,', function () {
 
   beforeAll(function (done) {
-    Promise.all(setGeneralInteractions(provider)).then(done, done.fail)
+    const promises = []
+    promises.push(setGeneralInteractions(provider))
+    // a request to GET attributes of an app
+    const attributes = {
+      attr1: {
+        attrExists: true,
+        value: 'value1'
+      },
+      'attr+plus space': {
+        attrExists: true,
+        value: 'value+plus space and/slash'
+      },
+      属性1: {
+        attrExists: true,
+        value: '值对1'
+      },
+      'attr ': {
+        attrExists: false,
+        value: ''
+      },
+      'attr+plus space ': {
+        attrExists: false,
+        value: ''
+      },
+      '属性1 ': {
+        attrExists: false,
+        value: ''
+      },
+      '': {
+        attrExists: false,
+        value: {
+          attr1: 'value1',
+          'attr+plus space': 'value+plus space and/slash',
+          属性1: '值对1'
+        }
+      },
+      'attr1-no-value': {
+        attrExists: true,
+        value: ''
+      },
+      'attr+plus space-no-value': {
+        attrExists: true,
+        value: ''
+      },
+      '属性1-no-value': {
+        attrExists: true,
+        value: ''
+      }
+    }
+    for (const attribute in attributes) {
+      // default no data
+      let data = ' <data/>\n'
+
+      // no attributes specified, return all attributes
+      if (attribute === '') {
+        data = ' <data>\n'
+        for (const [key, value] of Object.entries(attributes[attribute].value)) {
+          data = data +
+            '  <element>\n' +
+            '   <key>' + utf8.encode(key) + '</key>\n' +
+            '   <app>someAppName</app>\n' +
+            '   <value>' + utf8.encode(value) + '</value>\n' +
+            '  </element>\n'
+        }
+        data = data + ' </data>'
+      } else if (attributes[attribute].attrExists === true) {
+        // attribute exists
+        data = ' <data>\n' +
+          '  <element>\n' +
+          '   <key>' + utf8.encode(attribute) + '</key>\n' +
+          '   <app>someAppName</app>\n' +
+          '   <value>' + utf8.encode(attributes[attribute].value) + '</value>\n' +
+          '  </element>\n' +
+          ' </data>\n'
+      }
+      promises.push(provider.addInteraction({
+        uponReceiving: 'a request to GET ' + attribute + '-attribute of an app',
+        withRequest: {
+          method: 'GET',
+          path: Pact.Matchers.term({
+            matcher: '.*\\/ocs\\/v1\\.php\\/privatedata\\/getattribute\\/someAppName\\/?' + encodeURIComponent(utf8.encode(attribute)) + '$',
+            generate: '/ocs/v1.php/privatedata/getattribute/someAppName/' + encodeURIComponent(utf8.encode(attribute))
+          }),
+          headers: validAuthHeaders
+        },
+        willRespondWith: {
+          status: 200,
+          headers: xmlResponseHeaders,
+          body: '<?xml version="1.0"?>\n' +
+            '<ocs>\n' +
+            ocsMeta('ok', '100') +
+            data +
+            '</ocs>'
```

Comment:
We can define this into a constant to make the code DRY as it is repeated elsewhere as well.

Reply:
fixed